cloudinit/config/cc_mounts.py

```diff
@@ -223,42 +223,86 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
     return size
 
 
+def get_fstype(path):
+    """Return the type of the filesystem that path is on.
+    If path is a mount point itself, then the type of the filesystem mounted
+    on path rather than its parent dir."""
+    return util.get_mount_info(path)[1]
+
+
+def create_swapfile(fname, size):
+    """Size is in MiB."""
+
+    errmsg = "Failed to create swapfile '%s' of size %dMB via %s: %s"
+
+    def unlink_f(fname):
+        if os.path.exists(fname):
+            os.unlink(fname)
+
+    def create_swap(fname, size, method):
+        LOG.debug("Creating swapfile in '%s' on fstype '%s' using '%s'",
+                  fname, fstype, method)
+
+        if method == "fallocate":
+            cmd = ['fallocate', '-l', '%dM' % size, fname]
+        elif method == "dd":
+            cmd = ['dd', 'if=/dev/zero', 'of=%s' % fname, 'bs=1M',
+                   'count=%d' % size]
+
+        try:
+            util.subp(cmd, capture=True)
+        except util.ProcessExecutionError as e:
+            LOG.warning(errmsg, fname, size, method, e)
+            unlink_f(fname)
+
+    swap_dir = os.path.dirname(fname)
+    util.ensure_dir(swap_dir)
+
+    fstype = get_fstype(swap_dir)
+
+    if fstype in ("xfs", "btrfs"):
+        create_swap(fname, size, "dd")
+    else:
+        try:
+            create_swap(fname, size, "fallocate")
+        except util.ProcessExecutionError as e:
+            LOG.warning(errmsg, fname, size, "dd", e)
+            LOG.warning("Will attempt with dd.")
+            create_swap(fname, size, "dd")
+
+    try:
+        os.chmod(fname, 0o600)
```

Comment:
os.chmod isn't going to raise a ProcessExecutionError, so we can probably move that up just above  this try scope. It also should probably be util.chmod instead, which accounts for Selinux support too.